### PR TITLE
feat: add terramate experimental get-config-value

### PR DESCRIPTION
**Pull request by i4ki**
Implements the `experimental get-config-value <var accessor>` command:

Usage:

```
$ cat globals.tm
globals {
	hello = "hello world"
	values = ["a", "b", "c", "d"]
	obj = {
		level1 = {
			level2 = {
				val = "test"
			}
		}
	}
}

$ terramate experimental get-config-value global.hello
hello world
$ terramate experimental get-config-value --as-json global.hello
"hello world"
$ terramate experimental get-config-value global.values
["a", "b", "c", "d"]
$ terramate experimental get-config-value global.obj.level1.level2.val
test
$ terramate experimental get-config-value global.obj.level1.level2
{
  val = "test"
}
$ terramate experimental get-config-value global.obj
{
  level1 = {
    level2 = {
      val = "test"
    }
  }
}
$ terramate experimental get-config-value --as-json global.obj
{"level1":{"level2":{"val":"test"}}}
```

### CODE REVIEW

**Comment by waxb** [comment]
love it

**Review by zied-elouaer — APPROVED** [review]

File: cmd/terramate/cli/cli.go [hunk 1]
```diff
@@ -1215,6 +1226,64 @@ func (c *cli) partialEval() {
 	}
 }
 
+func (c *cli) getConfigValue() {
+	logger := log.With().
+		Str("action", "cli.GetConfigValue()").
```
Comment (katcipis) [inline cmd/terramate/cli/cli.go:1231]:
```suggestion
		Str("action", "cli.getConfigValue()").
```

File: cmd/terramate/cli/cli.go [hunk 2]
```diff
@@ -1215,6 +1226,64 @@ func (c *cli) partialEval() {
 	}
 }
 
+func (c *cli) getConfigValue() {
```
Comment (katcipis) [inline cmd/terramate/cli/cli.go:1229]: This is quite similar to cli.eval from the other PR, so I imagine it will be extracted in some way no ? (or Im missing some difference)

File: cmd/terramate/e2etests/exp_eval_test.go [hunk 3]
```diff
@@ -223,4 +223,117 @@ func TestExpEval(t *testing.T) {
...
+			},
+		},
+		{
+			name: "unsupported rootname",
+			expr: `local.value`,
+			want: runExpected{
+				Status:      1,
+				StderrRegex: "only terramate and global variables are supported",
+			},
+		},
```
Comment (katcipis) [inline cmd/terramate/e2etests/exp_eval_test.go:266]: for completeness maybe test when accessing an undefined metadata/global.

File: cmd/terramate/e2etests/exp_vendor_test.go [hunk 4]
```diff
@@ -132,6 +135,104 @@ func TestVendorModuleRecursive1DependencyIsPatched(t *testing.T) {
 	assert.EqualStrings(t, "", string(got))
 }
 
+func TestModVendorRecursiveMustPatchAlreadyVendoredModules(t *testing.T) {
```
Comment (katcipis) [inline cmd/terramate/e2etests/exp_vendor_test.go:138]: this seems unrelated to the PR no ? :thinking:

File: cmd/terramate/e2etests/exp_vendor_test.go [hunk 5]
```diff
@@ -132,6 +135,104 @@ func TestVendorModuleRecursive1DependencyIsPatched(t *testing.T) {
...
+	// This reproduces the bug below:
+	//   - ModA depends on ModZ
+	//   - ModB depends on ModZ
+	// executing:
+	//   - tm vendor download ModA <ref>
+	//   - tm vendor download ModB <ref>
+	// will patch the ModA files to use ../modules/ModZ/...
+	// but will not for ModB because ModZ is already vendored.
```
Comment (katcipis) [inline cmd/terramate/e2etests/exp_vendor_test.go:146]: Can't we test this on the modvendor package instead of the e2etest ? the behavior/bug doesn't seem to be related to the CLI and could be tested directly on modvendor.Vendor no ? :thinking: (we get coverage + test doesn't depend on experimental CLI)

File: cmd/terramate/e2etests/exp_vendor_test.go [hunk 6]
```diff
@@ -132,6 +135,104 @@ func TestVendorModuleRecursive1DependencyIsPatched(t *testing.T) {
...
+	s.RootEntry().CreateConfig(Terramate(
+		Config(),
+	).String())
+	tmcli := newCLI(t, s.RootDir())
+	res := tmcli.run("experimental", "vendor", "download", modA, "main")
+	assertRunResult(t, res, runExpected{IgnoreStdout: true})
+	res = tmcli.run("experimental", "vendor", "download", modB, "main")
+	assertRunResult(t, res, runExpected{IgnoreStdout: true})
+	res = tmcli.run("experimental", "vendor", "download", modC, "main")
+	assertRunResult(t, res, runExpected{IgnoreStdout: true})
```
Comment (katcipis) [inline cmd/terramate/e2etests/exp_vendor_test.go:192]: Here instead of running terramate we could just use modvendor.Vendor (and move the test to modvendor), the rest would remain the same I think.

File: test/fs.go [hunk 7]
```diff
@@ -59,3 +59,13 @@ func AssertFileEquals(t *testing.T, filepath1, filepath2 string) {
 		t.Fatalf("-(%s) +(%s):\n%s", filepath1, filepath2, diff)
 	}
 }
+
+// AssertFileContentEquals asserts that file fname has the content of want.
+// It assumes the file content is a unicode string.
+func AssertFileContentEquals(t *testing.T, fname string, want string) {
+	t.Helper()
+	got := ReadFile(t, filepath.Dir(fname), filepath.Base(fname))
+	if diff := cmp.Diff(string(got), string(want)); diff != "" {
```
Comment (katcipis) [inline test/fs.go:68]:
```suggestion
	if diff := cmp.Diff(string(got), want); diff != "" {
```

File: tf/mod_source_parse.go [hunk 8]
```diff
@@ -115,7 +115,15 @@ func ParseSource(modsource string) (Source, error) {
 		rawURL := strings.TrimPrefix(modsource, "git::")
 		u, err := url.Parse(rawURL)
 		if err != nil {
-			return Source{}, errors.E(ErrInvalidModSrc, "invalid URL inside %s", modsource)
+			return Source{}, errors.E(ErrInvalidModSrc, modsource)
+		}
+
+		if u.Path == "" {
```
Comment (katcipis) [inline tf/mod_source_parse.go:121]: Don't we have a test for this ? (usually it is quite easy to test the parsing scenarios here)

**Review by katcipis — APPROVED** [review]
LGTM + comments

**Review by katcipis — DISMISSED** [review]

File: cmd/terramate/e2etests/exp_eval_test.go (hunk 3 above)
Comment (i4ki) [reply cmd/terramate/e2etests/exp_eval_test.go:276]: added

File: cmd/terramate/cli/cli.go (hunk 2 above)
Comment (i4ki) [reply cmd/terramate/cli/cli.go:1229]: there's some differences hard to extract, like this one works on the var traversals and serialization is different (has --as-json also).

File: cmd/terramate/e2etests/exp_vendor_test.go (hunk 4 above)
Comment (i4ki) [reply cmd/terramate/e2etests/exp_vendor_test.go]: this PR has nothing to do with vendoring, this was showing here probably because the branch wasn't updated with the main branch...

File: cmd/terramate/e2etests/exp_vendor_test.go (hunk 5 above)
Comment (i4ki) [reply cmd/terramate/e2etests/exp_vendor_test.go]: the same for all those other comments.

File: cmd/terramate/e2etests/exp_vendor_test.go (hunk 4 above)
Comment (katcipis) [reply cmd/terramate/e2etests/exp_vendor_test.go]: ahhh OK :laughing:

**Review by katcipis — APPROVED** [review]
:rocket: :rocket: :rocket:
